conans/client/client_cache.py

```diff
@@ -4,18 +4,43 @@
 from conans.client.conf import ConanClientConfigParser, default_client_conf, default_settings_yml
 from conans.model.values import Values
 from conans.client.detect import detect_defaults_settings
-from conans.model.ref import ConanFileReference
+from conans.model.ref import ConanFileReference, PackageReference
 from conans.model.manifest import FileTreeManifest
 from conans.paths import SimplePaths, CONANINFO
 from genericpath import isdir
 from conans.model.info import ConanInfo
+from conans.errors import ConanException
 
 
 CONAN_CONF = 'conan.conf'
 CONAN_SETTINGS = "settings.yml"
 LOCALDB = ".conan.db"
 REGISTRY = "registry.txt"
 PROFILES_FOLDER = "profiles"
+_CACHED_BUILD_ID = "_conn_b_id"  # no collision with existing user vars
+
+
+def has_build_id(conanfile):
+    return hasattr(conanfile, _CACHED_BUILD_ID)
+
+
+def build_id(conanfile):
+    if hasattr(conanfile, "build_id"):
+        # Caching build_id, to avoid recomputing it
+        if not hasattr(conanfile, _CACHED_BUILD_ID):
+            # construct new ConanInfo
+            build_id_info = conanfile.info.copy()
+            conanfile.info_build = build_id_info
+            # effectively call the user function to change the package values
+            try:
+                conanfile.build_id()
+            except Exception as e:
+                raise ConanException("Error in 'build_id()': %s" % str(e))
+            # compute modified ID
+            new_id = build_id_info.package_id()
+            setattr(conanfile, _CACHED_BUILD_ID, new_id)
+        return new_id
```

Comment:
This is why I used a private, complex name for _CACHED_BUILD_ID. Anyway, not necessary to cache it anymore.